File: CGStudio/PGCore/Math/Primitive3d/Space3d.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace PG.Core.Math
{
    public class Space3d
    {
        public void Add(IVolume3d volume)
        {
            this.volumes.Add(volume);
        }

        public bool IsInside(Vector3d position)
        {
            foreach(var v in volumes)
            {
                if(!v.IsInside(position))
                {
                    return false;
                }
            }
            return true;
        }

        public IEnumerable<IVolume3d> Volumes { get { return volumes; } }

        private readonly List<IVolume3d> volumes = new List<IVolume3d>();
    }
}

File: Crystal/Math/Space3d.cpp
#include "Space3d.h"

using namespace Crystal::Math;

void Space3d::add(IVolume3dd* volume)
{
	this->volumes.push_back(volume);
}

bool Space3d::isInside(const Vector3dd& position) const
{
	for (auto v : volumes) {
		if (!v->isInside(position)) {
			return false;
		}
	}
	return true;
}
